src/poly/cse.rs

```diff
@@ -0,0 +1,83 @@
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+
+use crate::field::Field;
+
+use std::{collections::HashSet, fmt::Debug, hash::Hash};
+
+use super::{Expr, HashResult, VarAssignments};
+
+/// Common Subexpression Elimination - takes a collection of expr
+pub fn cse<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
+    exprs: Vec<Expr<F, V, ()>>,
+) -> Vec<Expr<F, V, HashResult>> {
+    // get the key set for the assignments
+    let mut keys: HashSet<V> = HashSet::new();
+
+    for expr in exprs.iter() {
+        // get the assignments for the expression
+        let expr_queries = expr.get_queries();
+        // add the assignments to the key set
+        keys.extend(expr_queries);
+    }
+    // generate random point in the field set
+    let _assignments = generate_random_assignment::<F, V>(keys);
+
+    // hash all the functions using the point for the assignments
+
+    // check if the hash is already in the hash table
+    // if it is, then replace it by a reference to the other one
+
+    // return the expressions hashed and optimized
+
+    todo!()
+}
+
+fn generate_random_assignment<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
+    keys: HashSet<V>,
+) -> VarAssignments<F, V> {
+    let mut rng = ChaCha20Rng::seed_from_u64(0);
+
+    let mut assignments = VarAssignments::new();
+
+    for key in keys.iter() {
```

Comment:
For forward signals we should add one random for the rotation 0 and another for rotation 1

Reply:
Why shouldn't they use the same assignment? Because in the end they are like the same value being used in two different PI's right? Or maybe I'm just not making any sense 😅

Reply:
The assignments are random, there show me two random assignations to each forward, one for this step and another for the next. Because there can be queries to the next step (Queriable::Forward(_, true))